### ISSUE TITLE
Focus is lost after selecting option from "Select Menu" dropdown

### ISSUE THREAD

**Issue by tellthemachines** [WordPress/gutenberg#38169]
### Description

In the Navigation Block toolbar, the "Select Menu" button opens a dropdown from where a different menu can be selected. Selecting a menu from that dropdown causes the dropdown to close, but when it closes keyboard focus is reset to the document body. 

Focus should instead go back to the "Select Menu" button in the toolbar, or perhaps to the block itself.

Possibly due to https://github.com/WordPress/gutenberg/issues/39044

**Edit**: Now it only happens when creating a new menu or importing from a classic one.

### Step-by-step reproduction instructions

1. Create a Navigation block, and make sure the site has more than one navigation menu already.
2. In the Navigation block toolbar, click "Select Menu" and choose a menu from the dropdown.
3. Note focus is no longer on the block when the dropdown closes.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by tellthemachines** [comment]
Looks like what's happening is the toolbar re-renders when the nav block contents are replaced, causing the button that focus should be returned to to disappear from the DOM.

**Comment by getdave** [comment]
FYI that this is being looked at in https://github.com/WordPress/gutenberg/pull/38907#discussion_r813555589.

We decided not to fix in that PR as this Issue is too complex.

**Comment by getdave** [comment]
Whoever works on a fix here should be sure to include an e2e test to ensure the `Select menu...` is not unmounted during loading.

**Comment by getdave** [comment]
I tested this by removing the `RecursionProvider` from the block. The focus loss issue still occurs. It's something else...

**Comment by alexstine** [comment]
This was a fairly complicated issue. I think this happens when the block changes in to loading state. I made a fix but it probably isn't pretty. Hopefully it may serve as a guide moving forward. I think this happens when the placeholder dynamically changes and focus is not forced. The block itself has some interesting focus issues even outside of this issue. 👎

**Comment by talldan** [comment]
> I tested this by removing the RecursionProvider from the block. The focus loss issue still occurs. It's something else...

@getdave That's not a very scientific test, because there can be more than one cause of a bug. That's exactly the case in the navigation block, lots of DOM elements are unmounted/remounted, and that's one cause.

In a past attempt to solve this I removed as much code that causes unmounting/remounting as I could and the Recursion Provider was still an issue.

**Comment by getdave** [comment]
> @getdave That's not a very scientific test, because there can be more than one cause of a bug.

I guess I didn't include my nuance here. What I meant to say was there doesn't appear to be a single cause here.

**Comment by draganescu** [comment]
I think  #42987 solved this by incorporating https://github.com/WordPress/gutenberg/pull/42956 but I am not confident enough to close the issue.

**Comment by cbravobernal** [comment]
Tested and is working fine!
